### DISCUSSION TITLE
Nested reusable workflows and matrices: parsing error

### DISCUSSION THREAD

**Discussion by trueneu** [community/community#32023] [Actions]
Hey.
I'm using 2-level nested re-usable workflows, so the main workflow M calls workflow A that calls another workflow B. M runs a job that forms a JSON that's passed as a string to A. A then has two jobs: first, JOB1, calls B with matrix strategy, reading matrix with `fromJson`; second, JOB2, has some explicit steps defined, also with matrix strategy from the exact same matrix. JOB2 `needs` JOB1.
Now the funny thing is, as soon as I introduce JOB1 into the workflow A, it fails claiming it can't render the strategy (specific details ommitted):
```
Error when evaluating 'strategy' for job 'JOB1'. path-to-yml (Line: X, Col: Y): Error parsing fromJson,path-to-yml (Line: X, Col: Y): Error reading JToken from JsonReader. Path '', line 0, position 0.,path-to-yml (Line: X, Col: Y): Unexpected type of value '', expected type: Mapping.
```

So from the error message it seems that the matrix string is empty and can't be parsed. However the debug output (just echoing the string) in a separate job shows it's a perfectly valid JSON. As soon as I comment out JOB1, JOB2 runs perfectly with literally no other changes. If I change the aforementioned debug job to also use the matrix strategy and make it needed for JOB2, everything runs once again. If I call workflow B directly from workflow M, it also works.

I can't really show a minimal reproducible example right now, but perhaps the description is enough. If it helps, I'll add it later.

UPD: please see the minimal reproducible example here: https://github.com/orgs/community/discussions/32023#discussioncomment-3561655

**Comment by airtower-luna** [comment]
> I can't really show a minimal reproducible example right now, but perhaps the description is enough.

Probably not, because the exact structure of both the JSON and how the matrix is initialized from is critical here. I've seen enough people use the approach, so most likely something is *slightly* off.

**Reply by trueneu** [reply]
@airtower-luna 

Please take a look here: https://github.com/trueneu/github-actions-nested-workflows

In case it's more convenient for the conversation, here are the files:

main.yml:
```
name: main workflow
on:
  push:
    branches:
      - '**'

jobs:
  matrix-gen:
    name: generate matrix
    runs-on: ubuntu-latest
    outputs:
      output-matrix: ${{ steps.generate.outputs.output-matrix }}
    steps:
      - run: |
          echo -n "::set-output name=output-matrix::{\"include\": [{\"key\": \"value1\"}, {\"key\": \"value2\"}]}"
        id: generate

  debug-output-main:
    runs-on: ubuntu-latest
    needs:
      - matrix-gen
    steps:
      - run: |
          echo ${{ needs.matrix-gen.outputs.output-matrix }}

  call-A:
    needs:
      - matrix-gen
    uses: ./.github/workflows/A.yml
    with:
      matrix-input: ${{ needs.matrix-gen.outputs.output-matrix }}

  call-B-from-main:
    uses: ./.github/workflows/B.yml
    needs:
      - matrix-gen
    strategy:
      matrix: ${{ fromJson(needs.matrix-gen.outputs.output-matrix) }}
    with:
      value: ${{ matrix.key }}

```

A.yml:
```
name: reusable workflow A
on:
  workflow_call:
    inputs:
      matrix-input:
        type: string
        required: true

jobs:
  debug-output-A:
    name: debug output
    runs-on: ubuntu-latest
    strategy:
      matrix: ${{ fromJson(inputs.matrix-input) }}
    steps:
      - run: |
          echo '${{ matrix.key }}'

  call-B-plain:
    uses: ./.github/workflows/B.yml
    with:
      value: 'arbitrary string'

  call-B-with-matrix:
    uses: ./.github/workflows/B.yml
    strategy:
      matrix: ${{ fromJson(inputs.matrix-input) }}
    with:
      value: ${{ matrix.key }}

```

B.yml:
```
name: reusable workflow B
on:
  workflow_call:
    inputs:
      value:
        type: string
        required: true

jobs:
  debug-output-B:
    name: debug output
    runs-on: ubuntu-latest
    steps:
      - run: |
          echo '${{ inputs.value }}'
```

**Reply by trueneu** [reply]
`call-B-from-main` will succeed, but `call-B-with-matrix` will fail to render the strategy.

**Reply by airtower-luna** [reply]
The only difference I can spot there is the use of the `inputs` context in `A.yml`, instead of `needs` in `main.yml`. I assume that means `inputs` isn't available for defining a matrix.

I'm not sure if that's intentional, the [documentation on `inputs` says](https://docs.github.com/en/actions/learn-github-actions/contexts#inputs-context):

> You can access this context from any job or step in a workflow.

Personally I'd consider `matrix` part of a job, but maybe they mean each matrix instance? They are separate jobs in the UI after all… :confused: 

You can try to report a bug, and maybe a staff member see will this thread.

**Reply by trueneu** [reply]
That would hold, but if only the other job using inputs to define the matrix wouldn't work either. It seems like an inconsistency or discrepancy of some sort. I will report a bug, thank you.

**Comment by trueneu** [comment]
For anybody interested, here's the ticket - https://support.github.com/ticket/personal/0/1776040

**Reply by airtower-luna** [reply]
That doesn't seem to be visible to the public, but I'd be curious to hear about the answer when you get one. :slightly_smiling_face:

**Reply by trueneu** [reply]
Oh, well. I was worried about "personal" part in the URL, but the ticket id is definitely global order. Well, the ticket itself doesn't have any new information anyway; as soon as I receive a definite reply, I'll post here.

**Reply by trueneu** [reply]
The problem was forwarded to the engineering team, the next reply is due to Sep, 14th. :)

**Reply by trueneu** [reply]
Got a reply from the GitHub support:

```
Hello Pavel,

Our engineering team has a fix developed for this issue behavior. Due to the fix involving a core part of the matrix and nested reusable workflow features, the team is still undergoing some testing before deploying it out broadly.

The root cause was determined to be due to the current function being used for getting execution context not accounting for ancestors - so in a nested reusable flow, calling a reusable with an input value for the matrix strategy results in the passed input value being null in the context data. Then, when the strategy element is evaluated, there is no input data present which leads to the empty json string error.

I will share an update by next week at the latest (October 14th) with an update regarding ETA on the fix. Please let me know if you have any additional questions or concerns!

Best,

Arthur
GitHub Support
```

**Reply by tdmax** [reply]
Hello, has there been any update by chance for this issue from Github Support?  Have they mentioned when they will be releasing the fix?

**Reply by daanteb** [reply]
I'm also interested if there are any updates?

**Reply by trueneu** [reply]
They're still rolling out fixes, as the initial one uncovered something else. Still waiting on this. No ETA at the moment.

**Reply by Dramelac** [reply]
Hello, same problem here, any update or public issue available (the ticket URL is unreachable) ?

**Reply by ShutdownRepo** [reply]
Also encountering this issue, very interested in any ETA and progress anyone could share on this matter 😄

**Reply by jasondamour** [reply]
Same here

**Reply by daanteb** [reply]
It seems that this issue is fixed, I retested our setup and it correctly created the matrix strategy. Could not find anything related in the change log though.

We're currently running runner version `2.301.1`

**Reply by ShutdownRepo** [reply]
> It seems that this issue is fixed, I retested our setup and it correctly created the matrix strategy. Could not find anything related in the change log though.
> 
> We're currently running runner version `2.301.1`

I'm not sure it is. I tried on my end and still encounter the same issues.

**Reply by ShutdownRepo** [reply]
@trueneu do you have any progress on your ticket?

**Reply by trueneu** [reply]
Hey, I was told it is fixed, the fixed is rolled out after some back and forth, and my test workflow now passes, you can take a look here - https://github.com/trueneu/github-actions-nested-workflows .
This is on a cloud runner, however, no idea how that translates to the self-hosted ones. I'd guess it maps 1-to-1 but who knows :D

Quote:
```
Our engineering team updated me that the fix has been rolled out for this behavior.

Please let us know if you have any additional questions or concerns, and thank you again for your patience with this process!
```

I don't have any means to reproduce it now, so if you still encounter the problem, feel free to open your own ticket and include my ticket as a reference (I suspect that, even if you can't open it, GitHub support still can)

Also, throughout the ticket I learned GHA has per-account feature flags; and as the fix initially was conflicting with some other GHA behaviour, the support has enabled the fix flag for my account specifically. It _could be_ that the general fix doesn't work for everybody because they've forgotten to actually enable it for everybody. Go figure :(

**Reply by ShutdownRepo** [reply]
Thank you @trueneu ! I have forked your showcase, and made the workflows run on self-hosted runners, and it all goes well. However, I have been battling with GHA for two days now trying to figure out why my own workflows don't work and it can only be one of two solutions.
1. I messed something up. But I showed the code to a few people, made quite some tests...
2. the most probable case: GHA is still having an issue, that probably differs a bit from the one you had, and it hasn't been acknowledged/fixed yet. 🤷

**Reply by ShutdownRepo** [reply]
I modified my workflow to remove everything that wasn't related to the failures, similarly to what @trueneu did in his original showcase: https://github.com/ShutdownRepo/github-actions-nested-workflows

I'm completely lost on what doesn't work. If anyone has an idea on what could be wrong I'd be very thankful.
~~In the mean time, I'll try to open a ticket a identify if this is on GHA's side~~

EDIT: it seems I cannot open a ticket and it's limited to enterprise accounts...

**Reply by trueneu** [reply]
> I modified my workflow to remove everything that wasn't related to the failures, similarly to what @trueneu did in his original showcase: https://github.com/ShutdownRepo/github-actions-nested-workflows
> 
> I'm completely lost on what doesn't work. If anyone has an idea on what could be wrong I'd be very thankful.
> ~~In the mean time, I'll try to open a ticket a identify if this is on GHA's side~~
> 
> EDIT: it seems I cannot open a ticket and it's limited to enterprise accounts...

Hey, I'll take a look in the next couple of days, maybe I can spot something.

**Reply by ChristopherHX** [reply]
> anyone has an idea on what could be wrong I'd be very thankful

I tracked down the potential issue in
https://github.com/ChristopherHX/test-github-actions-nested-workflows/tree/three-level-matrix/.github/workflows

- If you uncomment the dependency in workflow l1 or l2, Github Actions get's an internal error. (l0 and l3 are both working while referencing `needs` and have dependency)
- If you remove the context reference to `needs` from the matrix defintion it works again (some code seem to check for the `needs` context in the expresson)

_It's another  GitHub Actions reusable workflow bug ([my actions emulation utility has no problems with your workflow file](https://github.com/ChristopherHX/runner.server))_

Summary
`needs` means referencing the `needs` context in the matrix strategy expression. 
nesting level|`matrix` + `needs` + dependency|`matrix` + dependency|`matrix` + `needs` (= null)
---|---|---|---
0 - caller workflow|works|works|works
1 - reusable caller workflow|broken|works|works
2 - reusable caller workflow|broken|works|works
3 - reusable workflow|works|works|works

nesting level|`matrix` + `needs` + dependency|`matrix` + dependency|`matrix` + `needs` (= null)
---|---|---|---
0 - caller workflow|works|works|works
1 - reusable caller workflow|broken|works|works
2 - reusable workflow|works|works|works

**Reply by ShutdownRepo** [reply]
@ChristopherHX I'll be honest, I don't fully understand your answer. I'm not able to figure out how to apply your tests and conclusions to my showcase to make it work. Did you ? Or maybe I misunderstood your point

**Reply by ChristopherHX** [reply]
Your showcase seem to have no workaround (I'm with GitHub free, no enterprise user)

It doesn't seem to be possible to use `inputs` and `needs` within the matrix definition inside a reusable workflow to call another reusable workflow.

<Details>
<summary>It feels like that the matrix is evaluated twice</summary>

- once with invalid `inputs` and `needs` contexts (fromjson error)
- another time with valid `inputs` and `needs` contexts
- if both of the above cases evaluate matrices return different matrices => internal error - no error message

</details>

**Reply by trueneu** [reply]
@ShutdownRepo @ChristopherHX I had no time to dig deep into the problem myself, so I don't fully understand what's going on here, but you've mentioned you can't open a GitHub Support ticket as you're on free. If you'd write a ticket text, with full explanation of what's going on here, I'll be happy to file it for you and keep you updated on the progress here.

**Reply by ShutdownRepo** [reply]
Thank you @trueneu ! You'll find everything here: https://github.com/actions/runner/issues/2424
I don't know if it's more relevant to have an issue there or a ticket on your end, but if you feel it's worth a shot, everything needed to understand and replicate is in the issue I linked

**Reply by trueneu** [reply]
@ShutdownRepo no idea either to be honest, I'm pretty new to filing issues with GitHub :) Thanks for the link! I'll create one, we'll see.

**Reply by ShutdownRepo** [reply]
Thank you @trueneu 🙏

**Reply by trueneu** [reply]
@ShutdownRepo I was told that it's been reproduced and forwarded to the Engineering Team.

**Reply by trueneu** [reply]
Hey @ShutdownRepo , I've got the message that the behaviour was fixed, can you confirm?

**Reply by ShutdownRepo** [reply]
> Hey @ShutdownRepo , I've got the message that the behaviour was fixed, can you confirm?

Wow indeed, what a great news! Tested and confirmed https://github.com/ShutdownRepo/github-actions-nested-workflows/actions/runs/4383907692

**Reply by trueneu** [reply]
Great!
